frontend/catalyst/pennylane_extensions.py

```diff
@@ -2601,6 +2602,117 @@ def getArgsAsJAXArrays(self, flat_args):
         return jnpargs
 
 
+def pure_callback(callback_fn, result_type=None):
+    """Pure callback
+
+    A pure function is a function that:
+
+      1. Given the same arguments *args, the results will be the same each time the function is
+         called.
+      2. The function has no side effect.
+
+    A pure callback is a pure python function that can be executed by the python virtual machine.
+    This is in direct contrast to functions which get JIT compiled by Catalyst.
+
+    Using `pure_callback` allows a user to run python.
+    `pure_callback`s can be used via a decorator:
+
+    ```python
+    @pure_callback
+    def add_1(x) -> int:
+        return x + 1
+
+    @qjit
+    def context(x):
+        return add_1(x)
+
+    # Can also be used outside a JIT compiled context
+    two = add_1(1)
+    ```
+
+    It can also be used through a more functional syntax:
+
+
+    ```python
+    def add_1(x) -> int:
+        return x + 1
+
+    @qjit
+    def context(x):
+        return pure_callback(add_1, int)(x)
```

Comment:
Showcase the distinction since the type is passed explicitly?
```suggestion
    def add_1(x):
        return x + 1

    @qjit
    def context(x):
        return pure_callback(add_1, int)(x)
```